### PR TITLE
refactor: unify the developing and test environment

### PR DESCRIPTION
**Pull request by fengzixu**
Signed-off-by: fengzixu <hnustphoenix@gmail.com>

### Ⅰ. Describe what this PR did

Unify the developing and test environment according to the issue: https://github.com/dragonflyoss/Dragonfly/issues/260

1. refactor Makefile of Dragonfly.
2. use Docker to build Dragonfly other than local environment.
3. update CONTRIBUTING.md with the developing environment info.
4. update the config of Travis and circleci

### Ⅱ. Does this pull request fix one issue?

fixes: #260

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)

Just a Makefile.

### Ⅳ. Describe how to verify it

I have processed all commands in Makefile and compared behaviours of previous build file in the dir called `build`.

### Ⅴ. Special notes for reviews

Please help me check  whether behaviours are the same between Makefile in the root dir and build dir

### CODE REVIEW

**Comment by fengzixu** [comment]
It seems like that I need to repair ci. :)

**Comment by fengzixu** [comment]
@Starnop @lowzj @allencloud Please review this PR.

**Comment by fengzixu** [comment]
@lowzj Request review again.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	    -e GOOS=$(GOOS)                                                    \
+	    -e GOARCH=$(GOARCH)                                                \
+	    -e CGO_ENABLED=0                                                   \
+	    -w /go/src/$(PKG)                                                  \
+	    $(BUILD_IMAGE)                                                     \
+	    go install -v -pkgdir /go/pkg $(LDFLAGS_DFDAEMON) ./cmd/dfdaemon
+.PHONY: build-dfdaemon	
+
+build-supernode:
+	./hack/check-supernode.sh
```
Comment (Starnop) [inline Makefile]: `build-supernode` should no include the operation `check-supernode`, WDYT?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+uninstall:
+	@echo "Begin to uninstall dfget and dfdaemon."
+	./hack/install-client.sh uninstall
+.PHONY: uninstall 
+
+build-client: check-client build-dfget build-dfdaemon
+.PHONY: build-client
+
+build: check-client build-dfget build-dfdaemon build-supernode
```
Comment (Starnop) [inline Makefile]: make `check-supernode` before `build-supernode`, as mentioned above.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	    --rm                                                               \
+	    -ti                                                                \
+	    -u $$(id -u):$$(id -g)                                             \
+	    -v $$(pwd)/.go:/go                                                 \
+	    -v $$(pwd):/go/src/$(PKG)                                          \
+	    -v $$(pwd)/$(BUILD_PATH):/go/bin                                        \
+	    -v $$(pwd)/.cache:/.cache                                          \
+	    -e GOOS=$(GOOS)                                                    \
+	    -e GOARCH=$(GOARCH)                                                \
+	    -e CGO_ENABLED=0                                                   \
```
Comment (Starnop) [inline Makefile]: If I'm not mistaken, we need to set this value to `1`.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	    --rm                                                               \
+	    -ti                                                                \
+	    -u $$(id -u):$$(id -g)                                             \
+	    -v $$(pwd)/.go:/go                                                 \
+	    -v $$(pwd):/go/src/$(PKG)                                          \
+	    -v $$(pwd)/$(BUILD_PATH):/go/bin                                         \
+	    -v $$(pwd)/.cache:/.cache                                          \
+	    -e GOOS=$(GOOS)                                                    \
+	    -e GOARCH=$(GOARCH)                                                \
+	    -e CGO_ENABLED=0                                                   \
```
Comment (Starnop) [inline Makefile]: So do with above.

File: hack/build-supernode-image.sh [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+curDir=$(cd "$(dirname "$0")" && pwd)
+SUPERNODE_SOURCE_HOME="${curDir}/../src/supernode"
+
+buildDockerImage() {
+    cd "${SUPERNODE_SOURCE_HOME}" || return
+    mvn clean package -DskipTests docker:build
+}
+
+buildDockerImage "$@"
```
Comment (Starnop) [inline hack/build-supernode-image.sh]: Try to click the key `Enter`, and you might not get this weird symbol in the end. So do with others.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -139,6 +140,20 @@ No matter commit message, or commit content, we do take more emphasis on code re
 
 PR is the only way to make change to Dragonfly project files. To help reviewers better get your purpose, PR description could not be too detailed. We encourage contributors to follow the [PR template](./.github/PULL_REQUEST_TEMPLATE.md) to finish the pull request.
 
+### Developing Environment
+
+During the make contributions to Dragonfly project, some necessary information about the developing environment should be showed to contributors.
```
Comment (Starnop) [inline CONTRIBUTING.md]:
```suggestion
 During the make contributions to Dragonfly project, some necessary information about the developing environment should be showed to contributors.
```
It's weird, may be we can say that like this: 
As a contributor, we should reach an agreement on the version of the tools used in the development environment.  
FYI, WDYT?

File: Makefile (hunk 3 above)
Comment (fengzixu) [reply Makefile]: It is disabled by default when cross-compiling. And I don't find any reason to enable it.

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+PKG := github.com/dragonflyoss/Dragonfly
+SUPERNODE_SOURCE_HOME="${curDir}/../../src/supernode"
+BUILD_IMAGE ?= golang:1.10.4
+GOARCH := $$(go env GOARCH)
+GOOS := $$(go env GOOS)
+BUILD := $(shell git rev-parse HEAD)
+BUILD_PATH := release/${GOOS}_${GOARCH}
```
[the remark this answers is not in the document]
Comment (fengzixu) [reply Makefile]: It seems like that if we compile the dfget and dfdaemon in linux, it doesn't create `linux_amd64` dir. But it creates the `darwin_amd64` dir when we compile it in macOS. So whether we should specify the GOOS and GOARCH statically, such as GOOS=linux GOARCH=amd64

**Comment by fengzixu** [comment]
> @fengzixu One point, there are a lot of duplicate codes in the `build` and `hack` directories. Can we combine them?

sure

**Comment by fengzixu** [comment]
@lowzj @Starnop Review again.

**Comment by allencloud** [comment]
Thanks for your work. Merging...

**Pull request merged**